### DISCUSSION TITLE
Pagination is inconsistent for collections

### DISCUSSION THREAD

**Discussion by ThePlatinum** [laravel/framework#51085] [General] [answered]
When an api resource collection is paginated, the structure of the pagination is not consistent with if it were a query result.

From the doc: https://laravel.com/docs/11.x/pagination, pagination without resource collection does not wrap the meta information in a "metas" key as seen in this structure:
![image](https://github.com/laravel/framework/assets/30179079/317535ff-35a5-4ee4-836d-769ab7ecef76)

But, for resource collections, the structure is different, as the meta information about the number of pages are wrapped in "metas", even if the `withoutWrapping` method has been called.
![image](https://github.com/laravel/framework/assets/30179079/a3fa572f-383e-45b2-b6b5-67b23c05c208)

Does anyone know I can make the pagination of the resource collection have same structure as the regular pagination as I need to maintain consistency in my APIs.

I am thinking of customizing the pagination information as stated here: https://laravel.com/docs/11.x/eloquent-resources#customizing-the-pagination-information, but, I fear it will still keep the meta and that will be duplication of data on the responce.

**Comment by macropay-solutions** [comment]
Hi. You can use macros to overwrite the default behavior. 
https://github.com/laravel/framework/blob/11.x/src/Illuminate/Http/Resources/Json/PaginatedResourceResponse.php

```
    /**
     * Add the pagination information to the response.
     *
     * @param  \Illuminate\Http\Request  $request
     * @return array
     */
    protected function paginationInformation($request)
    {
        $paginated = $this->resource->resource->toArray();

        $default = [
            'links' => $this->paginationLinks($paginated),
            'meta' => $this->meta($paginated),
        ];

        if (method_exists($this->resource, 'paginationInformation') ||
            $this->resource->hasMacro('paginationInformation')) {
            return $this->resource->paginationInformation($request, $paginated, $default);
        }

        return $default;
    }
```

As you can see this is not macroable so you must check where it is called and see if it could be overwritten with custom logic.

PS. We created a restful crud lib to solve this: https://packagist.org/packages/macropay-solutions/laravel-crud-wizard-free

**Comment by ThePlatinum** [comment]
Thank you... At least now I see "How" Laravel does it, I will try to see how I can override it.

**Reply by macropay-solutions** [reply]
```
        if (method_exists($this->resource, 'paginationInformation')) {
            return $this->resource->paginationInformation($request, $paginated, $default);
        }
```
try to define this function in your resource.

as


```
    public function paginationInformation(\Illuminate\Http\Request $request, array $paginated, array $default): array
    {
        return array_merge(
           $paginated,
           $default['links'] ?? [],
           $default['meta'] ?? [],
        );
    }
```

**Reply by ThePlatinum** [reply]
it didn't work... 

I also can't override the method as the JsonResource class does not directly extend the PaginatedResourceResponse.php class... I am thinking a macro on the JsonResource will work... since it uses a trait (DelegatesToResource) that is Macroable

**Comment by ThePlatinum** [comment, marked as answer]
Solved by writing a macro on the JsonResource class, using the defination provided by @macropay-solutions

```
        JsonResource::macro('paginationInformation', function ($request, $paginated, $default) {
            return array_merge(
                $paginated,
                $default['links'] ?? [],
                $default['meta'] ?? [],
            );
        });
```

**Reply by macropay-solutions** [reply]
It is always a good idea to have return types and typed params. Your code lacks these.

**Reply by ThePlatinum** [reply]
Yeah... Sure I will write those just sharing what worked... I am a big fan of PSRs and PERs
